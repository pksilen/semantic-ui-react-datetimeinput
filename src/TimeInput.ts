// @flow

import React from 'react';
import type { Element } from 'react';
import PropTypes from 'prop-types';
import { Button } from 'semantic-ui-react';
// $FlowFixMe
import PropValidators from './PropValidators';
// $FlowFixMe
import Validators from './Validators';
// $FlowFixMe
import type { ButtonType, Props } from './Types';
// $FlowFixMe
import style from './style';
// $FlowFixMe
import NumberUtils from './NumberUtils';
// $FlowFixMe
import ButtonUtils from './ButtonUtils';

// noinspection JSUnusedGlobalSymbols
export default class NumberInput extends React.Component<Props, {}> {
  static propTypes = {
    value: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    allowEmptyValue: PropTypes.bool,
    buttonPlacement: PropTypes.oneOf(['right', 'leftAndRight']),
    className: PropTypes.string,
    defaultValue: PropTypes.number,
    id: PropTypes.string,
    minValue: PropTypes.number,
    maxValue: PropTypes.number,
    maxLength: PropTypes.number,
    placeholder: PropTypes.string,
    precision: PropTypes.number,
    showError: PropTypes.bool,
    size: PropTypes.oneOf(['mini', 'small', 'large', 'big', 'huge', 'massive']),
    stepAmount: PropTypes.number,
    valueType: PropTypes.oneOf(['integer', 'decimal'])
  };

  // noinspection MagicNumberJS
  static defaultProps = {
    allowEmptyValue: false,
    buttonPlacement: 'leftAndRight',
    className: undefined,
    defaultValue: undefined,
    id: undefined,
    minValue: 0,
    maxValue: 9999999999,
    maxLength: 10,
    placeholder: 'Enter a value',
    precision: 2,
    showError: false,
    size: 'small',
    stepAmount: 1,
    valueType: 'integer'
  };

  decrementValue = () => {
    const { minValue, onChange, precision, stepAmount, value, valueType } = this.props;
    const currentValue = NumberUtils.getParsedValue(value, valueType);

    if (Validators.isValidValue(currentValue, valueType)) {
      const newValue = currentValue - stepAmount;
      if (newValue >= minValue) {
        onChange(NumberUtils.getValueWithPrecisionAsString(newValue, valueType, precision));
      }
    }
  };

  incrementValue = () => {
    const { maxValue, onChange, precision, stepAmount, value, valueType } = this.props;
    const currentValue = NumberUtils.getParsedValue(value, valueType);

    if (Validators.isValidValue(currentValue, valueType)) {
      const newValue = currentValue + stepAmount;
      if (newValue <= maxValue) {
        onChange(NumberUtils.getValueWithPrecisionAsString(newValue, valueType, precision));
      }
    }
  };

  changeValue = ({ target: { value } }: SyntheticInputEvent<HTMLInputElement>) => {
    const { allowEmptyValue, maxValue, minValue, onChange, precision, valueType } = this.props;
    const newValue = NumberUtils.getParsedValue(value, valueType);

    if (Validators.isValidValue(newValue, valueType)) {
      if (newValue >= minValue && newValue <= maxValue) {
        onChange(NumberUtils.getValueWithPrecisionAsString(newValue, valueType, precision));
      }
    } else if (allowEmptyValue && !value) {
      onChange('');
    }
  };

  onInputBlur = () => {
    const { defaultValue, onChange, precision, value, valueType } = this.props;

    if (defaultValue !== undefined && !value) {
      onChange(NumberUtils.getValueWithPrecisionAsString(defaultValue, valueType, precision));
    }
  };

  getInputComponent = (): Element<*> => {
    const { buttonPlacement, maxLength, placeholder, showError, size, value } = this.props;
    const inputStyle = {
      ...style.common.input,
      ...style[buttonPlacement].input
    };

    return (
      <div className={`ui input ${size}${showError ? ' error' : ''}`}>
        <input
          type="text"
          style={inputStyle}
          maxLength={maxLength}
          placeholder={placeholder}
          value={value}
          onChange={this.changeValue}
          onBlur={this.onInputBlur}
        />
      </div>
    );
  };

  getButtonComponent = (buttonType: ButtonType): Element<*> => {
    const { buttonPlacement, size } = this.props;
    const buttonStyle = {
      ...style[buttonPlacement].button.base,
      ...style[buttonPlacement].button[buttonType]
    };

    return (
      <Button
        size={size}
        style={buttonStyle}
        icon={ButtonUtils.getButtonIconName(buttonType, buttonPlacement)}
        onClick={buttonType === 'increment' ? this.incrementValue : this.decrementValue}
        disabled={ButtonUtils.isDisabledButton(buttonType, this.props)}
      />
    );
  };

  render(): Element<*> {
    PropValidators.validatePropsInDevelopmentMode(this.props);
    const { buttonPlacement, className, id } = this.props;

    if (buttonPlacement === 'leftAndRight') {
      return (
        <div id={id} className={className} style={style.leftAndRight.outerDiv}>
          {this.getButtonComponent('decrement')}
          {this.getInputComponent()}
          {this.getButtonComponent('increment')}
        </div>
      );
    }

    return (
      <div id={id} className={className} style={style.right.outerDiv}>
        {this.getInputComponent()}
        <div style={style.right.button.div}>
          {this.getButtonComponent('increment')}
          {this.getButtonComponent('decrement')}
        </div>
      </div>
    );
  }
}
