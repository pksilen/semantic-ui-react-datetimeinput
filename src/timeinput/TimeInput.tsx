import * as React from 'react';
import * as PropTypes from 'prop-types';
import styleMap from '../styleMap';
import { OptionalTimeInputProps, TimeInputProps, TimeInputPropTypes } from './TimeInputProps';
import { TimeInputType } from '../types/TimeInputType';
import TimeInputPropsValidator from './propsvalidator/TimeInputPropsValidator';
import BaseInput from '../baseinput/BaseInput';
import { OptionalBaseInputProps } from '../baseinput/BaseInputProps';
import { ClickType } from '../types/ClickType';

// noinspection JSUnusedGlobalSymbols
export default class TimeInput extends BaseInput<TimeInputProps> {
  static propTypes: TimeInputPropTypes = {
    ...BaseInput.propTypes,
    hourSingleClickStepCount: PropTypes.number,
    hourDoubleClickStepCount: PropTypes.number,
    minuteSingleClickStepCount: PropTypes.number,
    minuteDoubleClickStepCount: PropTypes.number
  };

  static defaultProps: OptionalBaseInputProps & OptionalTimeInputProps = {
    ...BaseInput.defaultProps,
    hourSingleClickStepCount: 1,
    hourDoubleClickStepCount: 4,
    minuteSingleClickStepCount: 1,
    minuteDoubleClickStepCount: 5
  };

  timeoutIDMap: { [key: string]: number } = {
    'decrement hour': 0,
    'decrement minute': 0,
    'increment hour': 0,
    'increment minute': 0
  };

  clickToStepCountMap: { [key: string]: number } = {
    'hour single': this.props.hourSingleClickStepCount,
    'minute single': this.props.minuteSingleClickStepCount,
    'hour double': this.props.hourDoubleClickStepCount,
    'minute double': this.props.minuteDoubleClickStepCount
  };

  getTimeoutIDMap(): { [key: string]: number } {
    return this.timeoutIDMap;
  }

  getClickStepCount(inputType: TimeInputType, clickType: ClickType): number {
    return this.clickToStepCountMap[`${inputType} ${clickType}`];
  }

  render(): React.ReactElement {
    TimeInputPropsValidator.validatePropsInDevelopmentMode(this.props);

    const {
      onDateValueChange,
      dateValue,
      buttonPlacement,
      doubleClickDelayInMillis,
      focusColor,
      showTooltipDelayInMillis,
      showTooltips,
      size,
      valueType,
      hourSingleClickStepCount,
      hourDoubleClickStepCount,
      minuteSingleClickStepCount,
      minuteDoubleClickStepCount,
      ...restOfProps
    } = this.props;

    if (buttonPlacement === 'buttonsOutside') {
      return (
        <div {...restOfProps} style={styleMap.buttonsOutside.outerDiv}>
          {this.getButtonComponent('decrement', undefined, styleMap.buttonsOutside.button.decrement)}
          {this.getInputComponent('hour', styleMap.common.lightRightBorderColor)}
          {this.getInputComponent('minute', styleMap.common.noLeftBorder)}
          {this.getButtonComponent('increment', undefined, styleMap.buttonsOutside.button.increment)}
        </div>
      );
    }

    return (
      <div {...restOfProps} style={styleMap.buttonsInside.outerDiv}>
        {this.getInputComponent('hour', styleMap.buttonsInside.firstInput)}
        <div style={styleMap.buttonsInside.button.div}>
          {this.getButtonComponent('increment', 'hour', styleMap.buttonsInside.button.nonLast)}
          {this.getButtonComponent('decrement', 'hour', styleMap.buttonsInside.button.nonLast)}
        </div>
        {this.getInputComponent('minute', styleMap.buttonsInside.nonFirstInput)}
        <div style={styleMap.buttonsInside.button.div}>
          {this.getButtonComponent('increment', 'minute', styleMap.buttonsInside.button.lastIncrement)}
          {this.getButtonComponent('decrement', 'minute', styleMap.buttonsInside.button.lastDecrement)}
        </div>
      </div>
    );
  }
}
