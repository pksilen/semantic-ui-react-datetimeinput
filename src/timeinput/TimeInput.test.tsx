import * as React from 'react';
import PropTypes from 'prop-types';
import { mount, shallow as renderShallow } from 'enzyme';
import TimeInput from './TimeInput';

const date = new Date('2019-01-10T12:30:00.000');
let onDateValueChangeMock: (newValue: Date) => void;

beforeEach(() => {
  onDateValueChangeMock = jest.fn();
});

describe('propTypes', () => {

  it('should validate prop types', () => {
    // eslint-disable-next-line react/forbid-foreign-prop-types
    const { propTypes } = TimeInput;

    expect(propTypes.hourSingleClickStepCount).toBe(PropTypes.number);
    expect(propTypes.hourDoubleClickStepCount).toBe(PropTypes.number);
    expect(propTypes.minuteSingleClickStepCount).toBe(PropTypes.number);
    expect(propTypes.minuteDoubleClickStepCount).toBe(PropTypes.number);
  });
});

describe('defaultProps', () => {
  it('should set default prop values correctly', () => {
    const timeInputWrapper = mount(
      <TimeInput dateValue={new Date()} onDateValueChange={onDateValueChangeMock} />
    );

    expect(timeInputWrapper.props().hourSingleClickStepCount).toBe(1);
    expect(timeInputWrapper.props().hourDoubleClickStepCount).toBe(4);
    expect(timeInputWrapper.props().minuteSingleClickStepCount).toBe(1);
    expect(timeInputWrapper.props().minuteDoubleClickStepCount).toBe(5);
  });
});

describe('getTimeoutIDMap()', () => {
  it('should return correct timeoutID map', () => {
    const timeInput = new TimeInput({
      ...TimeInput.defaultProps,
      dateValue: new Date(),
      onDateValueChange: onDateValueChangeMock
    });

    const timeoutIDMap = timeInput.getTimeoutIDMap();

    expect(timeoutIDMap['decrement hour']).toBe(0);
    expect(timeoutIDMap['decrement minute']).toBe(0);
    expect(timeoutIDMap['decrement hour']).toBe(0);
    expect(timeoutIDMap['increment minute']).toBe(0);
  });
});

describe(' getClickStepCount()', () => {
  it('should return correct click step count', () => {
    const timeInput = new TimeInput({
      ...TimeInput.defaultProps,
      dateValue: new Date(),
      onDateValueChange: onDateValueChangeMock
    });

    expect(timeInput.getClickStepCount('hour', 'single')).toBe(timeInput.props.hourSingleClickStepCount);
    expect(timeInput.getClickStepCount('minute', 'single')).toBe(timeInput.props.minuteSingleClickStepCount);
    expect(timeInput.getClickStepCount('hour', 'double')).toBe(timeInput.props.hourDoubleClickStepCount);
    expect(timeInput.getClickStepCount('minute', 'double')).toBe(timeInput.props.minuteDoubleClickStepCount);
  });
});

describe('render()', () => {
  it('should render component correctly when buttonPlacement is "buttonsOutside"', () => {
    const timeInputWrapper = renderShallow(
      <TimeInput
        buttonPlacement="buttonsOutside"
        dateValue={date}
        onDateValueChange={onDateValueChangeMock}
      />
    );

    expect(timeInputWrapper).toMatchSnapshot();
  });

  it('should render component correctly when buttonPlacement is "buttonsInside"', () => {
    const timeInputWrapper = renderShallow(
      <TimeInput buttonPlacement="buttonsInside" dateValue={date} onDateValueChange={onDateValueChangeMock} />
    );

    expect(timeInputWrapper).toMatchSnapshot();
  });

  it('should render additional props to outside div', () => {
    const timeInputWrapper = mount(
      <TimeInput
        id="testId"
        className="testClassName"
        dateValue={date}
        onDateValueChange={onDateValueChangeMock}
      />
    );

    const outsideDivWrapper = timeInputWrapper.find('div').first();

    expect(outsideDivWrapper.props().id).toBe('testId');
    expect(outsideDivWrapper.props().className).toBe('testClassName');
  });
});
