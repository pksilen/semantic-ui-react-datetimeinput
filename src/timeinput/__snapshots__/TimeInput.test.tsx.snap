// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() should render component correctly when buttonPlacement is "buttonsInside" 1`] = `
<div
  style={
    {
      "display": "flex",
    }
  }
>
  <Popup
    content="Double-click to set to 0"
    disabled={false}
    mouseEnterDelay={500}
    mouseLeaveDelay={200}
    offset={0}
    on="hover"
    pinned={false}
    position="top left"
    trigger={
      <div
        className="ui input small"
      >
        <input
          maxLength={3}
          onChange={[Function]}
          onDoubleClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          style={
            {
              "borderBottomRightRadius": 0,
              "borderTopRightRadius": 0,
              "color": "#85b7d9",
              "fontWeight": "bold",
              "paddingLeft": "0.5em",
              "paddingRight": "0.5em",
              "textAlign": "right",
              "width": "2.5em",
            }
          }
          type="text"
          value="12"
        />
      </div>
    }
  />
  <div
    style={
      {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <Popup
      content="Double-click to increment by 4"
      disabled={false}
      mouseEnterDelay={500}
      mouseLeaveDelay={200}
      offset={0}
      on="hover"
      pinned={false}
      position="top left"
      trigger={
        <Button
          as="button"
          disabled={false}
          icon="caret up"
          onClick={[Function]}
          size="small"
          style={
            {
              "borderRadius": 0,
              "flex": "0 0 50%",
              "margin": 0,
              "maxHeight": "unset",
              "minHeight": "unset",
              "padding": "0 0.2em",
            }
          }
        />
      }
    />
    <Popup
      content="Double-click to decrement by 4"
      disabled={false}
      mouseEnterDelay={500}
      mouseLeaveDelay={200}
      offset={0}
      on="hover"
      pinned={false}
      position="top left"
      trigger={
        <Button
          as="button"
          disabled={false}
          icon="caret down"
          onClick={[Function]}
          size="small"
          style={
            {
              "borderRadius": 0,
              "flex": "0 0 50%",
              "margin": 0,
              "maxHeight": "unset",
              "minHeight": "unset",
              "padding": "0 0.2em",
            }
          }
        />
      }
    />
  </div>
  <Popup
    content="Double-click to set to 0"
    disabled={false}
    mouseEnterDelay={500}
    mouseLeaveDelay={200}
    offset={0}
    on="hover"
    pinned={false}
    position="top left"
    trigger={
      <div
        className="ui input small"
      >
        <input
          maxLength={3}
          onChange={[Function]}
          onDoubleClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          style={
            {
              "borderRadius": 0,
              "paddingLeft": "0.5em",
              "paddingRight": "0.5em",
              "textAlign": "right",
              "width": "2.5em",
            }
          }
          type="text"
          value="30"
        />
      </div>
    }
  />
  <div
    style={
      {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <Popup
      content="Double-click to increment by 5"
      disabled={false}
      mouseEnterDelay={500}
      mouseLeaveDelay={200}
      offset={0}
      on="hover"
      pinned={false}
      position="top left"
      trigger={
        <Button
          as="button"
          disabled={false}
          icon="caret up"
          onClick={[Function]}
          size="small"
          style={
            {
              "borderBottomLeftRadius": 0,
              "borderBottomRightRadius": 0,
              "borderTopLeftRadius": 0,
              "flex": "0 0 50%",
              "margin": 0,
              "maxHeight": "unset",
              "minHeight": "unset",
              "padding": "0 0.2em",
            }
          }
        />
      }
    />
    <Popup
      content="Double-click to decrement by 5"
      disabled={false}
      mouseEnterDelay={500}
      mouseLeaveDelay={200}
      offset={0}
      on="hover"
      pinned={false}
      position="top left"
      trigger={
        <Button
          as="button"
          disabled={false}
          icon="caret down"
          onClick={[Function]}
          size="small"
          style={
            {
              "borderBottomLeftRadius": 0,
              "borderTopLeftRadius": 0,
              "borderTopRightRadius": 0,
              "flex": "0 0 50%",
              "margin": 0,
              "maxHeight": "unset",
              "minHeight": "unset",
              "padding": "0 0.2em",
            }
          }
        />
      }
    />
  </div>
</div>
`;

exports[`render() should render component correctly when buttonPlacement is "buttonsOutside" 1`] = `
<div
  style={
    {
      "display": "flex",
    }
  }
>
  <Popup
    content="Double-click to decrement by 4"
    disabled={false}
    mouseEnterDelay={500}
    mouseLeaveDelay={200}
    offset={0}
    on="hover"
    pinned={false}
    position="top left"
    trigger={
      <Button
        as="button"
        disabled={false}
        icon="minus"
        onClick={[Function]}
        size="small"
        style={
          {
            "borderBottomRightRadius": 0,
            "borderTopRightRadius": 0,
            "margin": 0,
            "maxHeight": "unset",
            "minHeight": "unset",
          }
        }
      />
    }
  />
  <Popup
    content="Double-click to set to 0"
    disabled={false}
    mouseEnterDelay={500}
    mouseLeaveDelay={200}
    offset={0}
    on="hover"
    pinned={false}
    position="top left"
    trigger={
      <div
        className="ui input small"
      >
        <input
          maxLength={3}
          onChange={[Function]}
          onDoubleClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          style={
            {
              "borderColor": "rgba(34, 36, 38, .05)",
              "borderRadius": 0,
              "borderRightColor": "rgba(34, 36, 38, .05)",
              "color": "#85b7d9",
              "fontWeight": "bold",
              "paddingLeft": "0.5em",
              "paddingRight": "0.5em",
              "textAlign": "right",
              "width": "2.5em",
            }
          }
          type="text"
          value="12"
        />
      </div>
    }
  />
  <Popup
    content="Double-click to set to 0"
    disabled={false}
    mouseEnterDelay={500}
    mouseLeaveDelay={200}
    offset={0}
    on="hover"
    pinned={false}
    position="top left"
    trigger={
      <div
        className="ui input small"
      >
        <input
          maxLength={3}
          onChange={[Function]}
          onDoubleClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          style={
            {
              "borderColor": "rgba(34, 36, 38, .05)",
              "borderLeft": "none",
              "borderRadius": 0,
              "paddingLeft": "0.5em",
              "paddingRight": "0.5em",
              "textAlign": "right",
              "width": "2.5em",
            }
          }
          type="text"
          value="30"
        />
      </div>
    }
  />
  <Popup
    content="Double-click to increment by 4"
    disabled={false}
    mouseEnterDelay={500}
    mouseLeaveDelay={200}
    offset={0}
    on="hover"
    pinned={false}
    position="top left"
    trigger={
      <Button
        as="button"
        disabled={false}
        icon="plus"
        onClick={[Function]}
        size="small"
        style={
          {
            "borderBottomLeftRadius": 0,
            "borderTopLeftRadius": 0,
            "margin": 0,
            "maxHeight": "unset",
            "minHeight": "unset",
          }
        }
      />
    }
  />
</div>
`;
